qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,529 @@
+            mask: mask for the IP address pattern informing which parts of
+                  the IP address should be matched against. 0 means ignore,
+                  255 means match.
+        """
+        host_ip = ipaddress.ip_address(host)
+        network = ipaddress.ip_network("{}/{}".format(pattern, mask))
+        return host_ip in network
+
+    @_js_slot(str)
+    def dnsResolve(self, host):
+        """Resolve a DNS hostname.
+
+        Resolves the given DNS hostname into an IP address, and returns it
+        in the dot-separated format as a string.
+
+        Args:
+            host: hostname to resolve.
+        """
+        ips = QHostInfo.fromName(host)
+        if ips.error() != QHostInfo.NoError or len(ips.addresses()) == 0:
+            return ""
+        else:
+            return ips.addresses()[0].toString()
+
+    @_js_slot()
+    def myIpAddress(self):
+        """Get host IP address.
+
+        Return the server IP address of the machine Firefox is running on,
+        as a string in the dot-separated integer format.
+        """
+        return self.dnsResolve(QHostInfo.localHostName())
+
+    @_js_slot(str)
+    def dnsDomainLevels(self, host):
+        """Count number of DNS domain levels.
+
+        Return the number (integer) of DNS domain levels (number of dots)
+        in the hostname.
+
+        Args:
+            host: the hostname from the URL.
+        """
+        return host.count('.')
+
+    @_js_slot(str, str)
+    def shExpMatch(self, mstr, shexp):
+        """Test is string matches shell expression.
+
+        Return true if the string matches the specified shell expression.
+        Currently, the patterns are shell expressions, not regular
+        expressions.
+
+        Args:
+            mstr: Any string to compare (e.g. the URL, or the hostname).
+            shexp: a shell expression to compare against.
+        """
+        return fnmatch.fnmatchcase(mstr, shexp)
+
+    def _get_dt(self, gmt):
+        if self._fixed_dt is not None:
+            if gmt:
+                return self._fixed_dt.toUTC()
+            else:
+                return self._fixed_dt
+        else:
+            if gmt:
+                return QDateTime.currentDateTimeUtc()
+            else:
+                return QDateTime.currentDateTime()
+
+    @staticmethod
+    def _in_range(avals, vals, bvals, ranges):
+        """Test if a tuple is between two tuples in numerical order."""
+        carry_a = 0
+        carry_val = 0
+        carry_b = 0
+        for a, val, b, r in zip(avals, vals, bvals, ranges):
+            if a is not None:
+                a_c = a + carry_a
+                val_c = val + carry_val
+                b_c = b + carry_b
+                if a_c < val_c < b_c:
+                    return True
+                elif not a_c <= val_c <= b_c:
+                    return False
+                carry_a += a * r
+                carry_val += val * r
+                carry_b += b * r
+        return True
```

Comment:
Not exactly, I need to implement comparison with carry here. For example: is year 2016 and day 15 between 2014, 20 and 2017, 30? If we simply compare tuples the check would fail because 15 is not between 20 and 30. To do this correctly we need to first check years and then carry current year value to months and then to days -- this way we would compare 15 + 2016 * 12 * 31 with 20 + 2014 * 12 * 31 and 30 + 2017 * 12 * 31. BTW because we don't convert days back into years, we can just take the highest possible value for days in a month for this purpose.